### ISSUE TITLE
fastertransformer nan error

### ISSUE THREAD

**Issue by leerumor** [NVIDIA/DeepLearningExamples#169]
When using my own pretrained model, I got NaN from the eight transformer layer.
By reading "fastertransformer/cuda/open_attention.cu", I found that the softmax input is not protected (substract the max value), the right formula shoud be:
```
def softmax(x):
    """Compute softmax values for each sets of scores in x."""
    e_x = np.exp(x - np.max(x))
    return e_x / e_x.sum()
```
I'm trying to modify "softmax_kernel_v2", maybe someone can offer a better version

**Comment by IrishCoffee** [comment]
Good suggestion. I will update the code soon.

**Comment by leerumor** [comment]
> Good suggestion. I will update the code soon.

thanks very much!
I'm not very familiar with cuda, I just implemented a blockReduceMax as your blockReduceSum, but it doesn't work at all...can you give me some advice?
The below code always get "inf" for max
```
T warpReduceMax(T val)
{
  for(int mask = 16; mask > 0; mask >>= 1) 
    val = max(__shfl_xor_sync(FINAL_MASK, val, mask, 32), val);
  return val;
}

/* Calculate the max of all elements in a block */
template <typename T> 
  __inline__ __device__
T blockReduceMax(T val)
{
  static __shared__ T shared[32]; 
  int lane = threadIdx.x & 0x1f; 
  int wid = threadIdx.x >> 5;  

  val = warpReduceMax<T>(val);

  if(lane == 0) 
    shared[wid] = val;

  __syncthreads();

  //val = (threadIdx.x < (blockDim.x >> 5 )) ? shared[lane] : (T)(0.0f);
  val = (threadIdx.x < (blockDim.x >> 5 )) ? shared[lane] : (T)(-INFINITY);
  val = warpReduceMax<T>(val);
                                 
  return val;
}
```

**Comment by IrishCoffee** [comment]
Could you please provide a checkpoint file to us?

**Comment by leerumor** [comment]
> Could you please provide a checkpoint file to us?

Sorry, I don't think I can, it's a pretrained bert by my company.
there is no nan when using google's bert but I think this overflow risk should be prevented ;)
Unfortunately, the speed will be dropped:
V100 seq=128, fp16, (bsz=1 got 1.96->2.64), (bsz=32 got 15.26->16.55)
and the drop becomes bigger when I use tensorflow opkernel (bsz=1 +2ms, bsz=32 +10ms)

**Comment by IrishCoffee** [comment]
I don't quite understand what do you mean.

(1) Using the updated SoftMax kernel still causes nan now?

(2) The performance drop is?

**Comment by leerumor** [comment]
1. using the softmax I implemented above doesn't cause nan
2. as in my code, I added a reduceMax, the execution becomes longer

**Comment by IrishCoffee** [comment]
as in my code, I added a reduceMax, the execution becomes longer
->
Could you verify the performance drop using the C++ interface? /bin/transformer_fp16

**Comment by IrishCoffee** [comment]
Already sent to your icloud email address. Please take a look.

**Comment by IrishCoffee** [comment]
Fixed with updated code. Close the issue.
